lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4186,9 +4186,24 @@ bool SROA_Helper::LowerMemcpy(Value *V, DxilFieldAnnotation *annotation,
 
 /// MarkEmptyStructUsers - Add instruction related to Empty struct to DeadInsts.
 void SROA_Helper::MarkEmptyStructUsers(Value *V, SmallVector<Value *, 32> &DeadInsts) {
+  UndefValue *undef = UndefValue::get(V->getType());
   for (User *U : V->users()) {
-    MarkEmptyStructUsers(U, DeadInsts);
+    // Kill memcpy, set operands to undef for call and ret, and recurse
+    if (MemCpyInst *MC = dyn_cast<MemCpyInst>(U)) {
+      DeadInsts.emplace_back(MC);
+    } else if (CallInst *CI = dyn_cast<CallInst>(U)) {
+      for (auto &operand : CI->operands()) {
```

Comment:
1. You can avoid this loop by iterating over `V->uses()` and using `Use.set(UndefValue)`.
2. In any case, I don't think this is legit because you are modifying the use collection while iterating over it (indirectly, through users). Why isn't it enough to just do `V->replaceAllUsesWith(undef)`? I also find the recursion suspicious, and the `DeadInsts.emplace_back` below. What if you have this:

```
struct EmptyStruct {};
static int globalVariable = 1;
void foo(EmptyStruct s) { globalVariable = 42; }
int main() : OUT
{
  foo(EmptyStruct());
  return globalVariable;
}
```

Reply:
Will update to fix iteration issue, clarify recursion cases, and add assert if assumption for recursion is incorrect.